packages/app-builder-backend/src/resolvers/key-resolver.ts

```diff
@@ -0,0 +1,244 @@
+      idToken,
+    },
+  )
+}
+
+const undefinedIfNoId = <T>(idObj: (T & { id?: string }) | undefined) => {
+  return idObj && idObj.id ? idObj : undefined
+}
+
+const cleanMovement = (movement: KeyMovement): KeyMovement => ({
+  ...movement,
+  checkOutToNegotiator: undefinedIfNoId(movement.checkOutToNegotiator),
+  checkOutToContact: undefinedIfNoId(movement.checkOutToContact),
+  checkInNegotiator: undefinedIfNoId(movement.checkInNegotiator),
+  checkOutNegotiator: undefinedIfNoId(movement.checkOutNegotiator),
+})
+
+const APIKeyToKey = (key: APIKey): Key => ({
+  ...key,
+  type: key.type.value,
+  movements: key.movements.map(cleanMovement),
+})
+
+@Resolver(() => Key)
+export class KeyResolver {
+  constructor() {}
+
+  @Query(() => [Key])
+  async listPropertyKeys(@Arg('propertyId') propertyId: string, @Ctx() ctx: Context): Promise<Key[]> {
+    const { accessToken, idToken } = ctx
+    if (!accessToken || !idToken) {
+      throw new Error('unauthorized')
+    }
+    const keys = await getPropertyKeys(propertyId, accessToken, idToken)
+    return keys?.map(APIKeyToKey) || []
+  }
+
+  @Query(() => [KeyMovement])
+  async listPropertyKeyMovements(
+    @Arg('propertyId') propertyId: string,
+    @Arg('keyId') keyId: string,
+    @Ctx() ctx: Context,
+  ): Promise<KeyMovement[]> {
+    const { accessToken, idToken } = ctx
+    if (!accessToken || !idToken) {
+      throw new Error('unauthorized')
+    }
+    const key = await getPropertyKey(propertyId, keyId, accessToken, idToken)
+    return key.movements.map(cleanMovement)
+  }
+
+  @Query(() => KeyMovement)
+  async getPropertyKeyMovement(
+    @Arg('propertyId') propertyId: string,
+    @Arg('keyId') keyId: string,
+    @Arg('movementId') movementId: string,
+    @Ctx() ctx: Context,
+  ): Promise<KeyMovement> {
+    const { accessToken, idToken } = ctx
+    if (!accessToken || !idToken) {
+      throw new Error('unauthorized')
+    }
+    const key = await getPropertyKey(propertyId, keyId, accessToken, idToken)
+    const movement = key.movements.find((m) => m.id === movementId)
+    if (!movement) {
+      throw new Error('movement not found')
+    }
+    return cleanMovement(movement)
+  }
+
+  @Query(() => Key)
+  async getPropertyKey(
+    @Arg('propertyId') propertyId: string,
+    @Arg('keyId') keyId: string,
+    @Ctx() ctx: Context,
+  ): Promise<Key> {
+    const { accessToken, idToken } = ctx
+    if (!accessToken || !idToken) {
+      throw new Error('unauthorized')
+    }
+    const key = await getPropertyKey(propertyId, keyId, accessToken, idToken)
+    return APIKeyToKey(key)
+  }
+
+  @Mutation(() => Key)
+  async checkOutKey(
+    @Arg('propertyId') propertyId: string,
+    @Arg('keyId') keyId: string,
+    @Arg('movement', () => KeyMovementModelInput) movement: KeyMovementModelInput,
+    @Ctx() ctx: Context,
+  ): Promise<Key> {
+    const { accessToken, idToken } = ctx
+    if (!accessToken || !idToken) {
+      throw new Error('unauthorized')
+    }
```

Comment:
could you make this snippet reusable in any way? Just thinking if we changed the auth method it would take a lot of regex to find all the locations

Reply:
Might also be helpful to throw this as a http error that lambdas like? I'm not 100% on this but I know they do like an object with a status property `return { status: 401, message: "Unauthorised" 
 }`

Reply:
good call, done it the native way now with decorators 👍